src/link/interface.ts

```diff
@@ -0,0 +1,57 @@
+/* eslint-disable @typescript-eslint/no-unnecessary-type-constraint */
+/* eslint-disable no-use-before-define */
+import type { MultihashDigest } from '../hashes/interface'
+import type { MultibaseEncoder, MultibaseDecoder, Multibase } from '../bases/interface'
+import type { Phantom, ByteView } from '../block/interface'
+
+
+export type { MultihashDigest, MultibaseEncoder, MultibaseDecoder }
+export type Version = 0 | 1
+
+export type DAG_PB = 0x70
+export type SHA_256 = 0x12
+
+/**
+ * Represents an IPLD link to a specific data of type `T`.
+ *
+ * @template T - Logical type of the data being linked to.
+ * @template C - multicodec code corresponding to a codec linked data is encoded with
+ * @template A - multicodec code corresponding to the hashing algorithm of the CID
+ * @template V - CID version
+ */
+export interface Link<
+  Data extends unknown = unknown,
+  Format extends number = number,
+  Alg extends number = number,
+  V extends Version = 1
+  > extends Phantom<Data> {
+  readonly version: V
+  readonly code: Format
+  readonly multihash: MultihashDigest<Alg>
+
+  readonly byteOffset: number
+  readonly byteLength: number
+  readonly bytes: ByteView<Link<Data, Format, Alg, V>>
+
+
+  equals(other: unknown): other is Link<Data, Format, Alg, Version>
+
+  toString<Prefix extends string>(base?: MultibaseEncoder<Prefix>): ToString<Link<Data, Format, Alg, Version>, Prefix>
+  toJSON(): { version: V, code: Format, hash: Uint8Array }
+  link(): Link<Data, Format, Alg, V>
```

Comment:
can you talk about the choice of name here? would `asLink` be more consistent with what we've done to `CID`?

Reply:
this seems like it's a convenient way to downcast a concrete `CID` to a `Link`, but can't you just use a `CID` as if it's a `Link` wherever you need one, do you ever really need to cast it? what's the use-case here?

Reply:
Sorry I should have added some code comments here.

it is followup on our https://github.com/multiformats/js-multiformats/issues/185 discussion which I thought you wanted to include)

I did however since realized that I have some use cases where I’d want a class where `link():Promise<Link>` and it would be convenient to have equivalent on `Link` itself. I didn’t like `toLink` because method with such name returning a promise doesn’t seat well with me.

Reply:
oh yeah ... got popped from my memory stack, and seeing all of the additional functionality in context here makes me question this. I like the idea of `thing.link()`, but with so many other methods on this interface, it seems to diminish the value; we're pretty close to a `CID` already and I had in my head that a `Link` interface was going to be a minimal thing that you could just ask for the more full-featured thing from. But that probably doesn't make sense.